### ISSUE TITLE
Expression for binding the Context Value

### ISSUE THREAD

**Issue by ChristianGruen** [qt4cg/qtspecs#131]
**REVISED:** I’ve incorporated the feedback from the comments (thanks).

We have no expression yet to bind a value to the context value. Such an expression would be useful, among other things, to extend the focus function to sequences (`fn { . }`, see #129).

Here are 3 possible constructs for that, ordered by my personal preference:

### 1. Value Map Expression

```xquery
ValueExpr      ::=  ValidateExpr | ExtensionExpr | ValueMapExpr
ValueMapExpr   ::=  SimpleMapExpr ("~" SimpleMapExpr)*
SimpleMapExpr  ::=  PathExpr ("!" PathExpr)*

(: Example :)
//flower ~ (count(.) || ' flowers: ' || string-join(name, ', '))
```

The expression would be similar to the simple map expression (which we could rename to item map expression). The following equivalents would then exist for simple FLWOR expressions:

```xquery
for $i in (1 to 5) return string($i)  ≍  (1 to 5) ! string(.)
let $i := (1 to 5) return count($i)   ≍  (1 to 5) ~ count(.)
```

`fn { E }` could be rewritten to `fn($c) { $c ~ E }`.

### 2. Context Value Declaration

```xquery
ContextExpr  ::=  "context" "{" Expr "}" EnclosedExpr

(: Example :)
context { //flower } {
  count(.) || ' flowers: ' || string-join(name, ', ')
}
```

The result of the first expression defines the context value, the second expression can reference the context.

`fn { E }` could be rewritten to `fn($c) { context { $c } { E } }`.

### 3. Enhanced FLWOR expression (for the sake of completion)

Similar to variables, the dot could be used to bind and reference the context:

```xquery
LetBinding  ::=  ("." | ("$" VarName)) TypeDeclaration? ":=" ExprSingle
ForBinding  ::=  ("." | ("$" VarName)) TypeDeclaration? AllowingEmpty? PositionalVar? "in" ExprSingle

(: Example :)
let . := //flower
return count(.) || ' flowers: ' || string-join(name, ', ')
```

`fn { E }` could be rewritten to `fn($c) { let . := $c return E }`.

### Assessment

* The first solution looks most appealing to me. I like the analogy with the existing syntax for single items.
* We could choose the second solution if we believe that the expression will be rarely used.
* I‘ve backed away from the third solution; I think it would be too pervasive.

**Comment by graydon2014** [comment]
I agree that the first solution seems more promising.  Making `.` assignable seems like a way to introduce excessively subtle errors.

Would
```
declare function local:flowerThings($in as xs:string) as element()* {
    context { db:open($in) } {
     //*[@isFlower]
  }
}
```
work/make sense?

**Comment by ChristianGruen** [comment]
Yes, it would: The result of the function call `db:open` would be bound as context value, and the path expression would be evaluated on that context.

**Comment by graydon2014** [comment]
That seems like a strong nice to have.  So would
```
<xsl:context select="_expression_">
  ...sequence constructor
</xsl:context>
```
I recognize that `xsl:for-each` will change the context item in XSLT but I think it'd be clearer to be able to do it specifically.

**Comment by ChristianGruen** [comment]
One more suggestion:

### 3. Value Map Expression

```xquery
ValueExpr  ::=  ValidateExpr | ExtensionExpr | ValueMapExpr
ValueMapExpr  ::=  SimpleMapExpr ("\" SimpleMapExpr)*
SimpleMapExpr  ::=  PathExpr ("!" PathExpr)*

(: Example :)
//flower \ (count(.) || ' flowers: ' || string-join(./name, ', '))
}
```

The expression would be very similar to the simple map expression. The following equivalent representations would then exist for simple FLWOR expressions:

```xquery
for $i in (1 to 5) return string($i)   ≍   (1 to 5) ! string(.)   ≍   (1 to 5) -> string#1
let $i := (1 to 5) return count($i)    ≍   (1 to 5) \ count(.)    ≍   (1 to 5) => count#1
```

A context-binding fat arrow expression `U => { E }` would be equivalent to `U \ E`.

The range of remaining ASCII single-character symbols is decreasing. A two-character symbol may thus be more appropriate.

**Comment by rhdunn** [comment]
With the proposed `U \ E` syntax, I'm trying to understand what is different to just using `=>`. I'm not directly opposed to the syntax, just trying to understand where that fits within the language.

I like the other two proposals (and the equivalent XSLT equivalents).

In terms of readability, I actually find the FLWOR version more readable. Especially given that `.` has a well-defined meaning.

**Comment by rhdunn** [comment]
Note that this ties into my proposal/analysis for supporting optional parameters (https://github.com/qt4cg/qtspecs/issues/64). For example:

    declare function fn:string($arg as item()? := .) as xs:string;

That would be consistent with the FLWOR let binding -- e.g. `let $value := . return $value`.

**Comment by ChristianGruen** [comment]
> I'm trying to understand what is different to just using `=>`.

@rhdunn I try to answer this by comparing this to the simple map expression: If we had had `X ->{ E }` from the very beginning, there would have been no need for `X ! E`.

But thanks for your observation! It helps me to conceive that we could also regard the fat sequence-binding arrow `=> { ... }` as **The Expression for binding the Context Value**, and it would not require an additional equivalent writing (→ #129, Example 2).

**Comment by ChristianGruen** [comment]
### 4. Value Map Expression

If the rules for the thin and fat arrow are unified…

```
ArrowExpr    ::=  UnaryExpr ( ("->" | "=>") ArrowTarget )*
ArrowTarget  ::=  (ArrowStaticFunction ArgumentList) |
                  (ArrowDynamicFunction PositionalArgumentList) | EnclosedExpr
```

…and if the fat arrow is added as inline function operator for binding sequences…

```
InlineFunctionExpr  ::=  Annotation* (("function" FunctionSignature) |
                         (("->" | "=>") FunctionSignature?)) FunctionBody
```

…`=> { E }` could then be expanded to `function($x as item()*) as item()* { $x => { E } }`.

**Comment by graydon2014** [comment]
Definitely prefer `=>` to adding another operator.

This all seems like a good idea from this corner.

**Comment by michaelhkay** [comment]
I find the syntax

```
context { //flower } {
  count(.) || ' flowers: ' || string-join(./name, ', ')
}
```
confusing. Specifically, what happens if `//flower` selects (a) a singleton, (b) an empty sequence, (c) a multi-item sequence? From the prose description, it looks as if the intent is to bind to a singleton, suggesting that (b) and (c) are errors.

But what does this give us that "!" doesn't? We can surely write 

`//flower ! (count(.) || ' flowers: ' || string-join(./name, ', '))`

or if we want to insist on it being a singleton

`exactly-one(//flower) ! (count(.) || ' flowers: ' || string-join(./name, ', '))`

**Comment by ChristianGruen** [comment]
I agree that `!` works nice to bind single items. All proposals have been designed with the intention to bind sequences of arbitrary length, which is why I was talking about a »context value« instead of »context item«. It probably makes sense to read #129 first (I hope).

**Comment by michaelhkay** [comment]
We seem to have overlooked that the current draft XPath specification includes proposed syntax for this feature: see the "WithExpr" in section 4.1.

**Comment by ChristianGruen** [comment]
If I read this correctly, the current »WithExpr« has been proposed to set namespaces. Do you mean that it could also be used/extended to bind sequences to the (yet to be formalized) »context value«?

**Comment by michaelhkay** [comment]
Sorry, I put my comment against the wrong issue.

**Comment by ChristianGruen** [comment]
I’ve revised the original comment. The value map expression, as was proposed later on in this thread, looks most promising to me.

**Comment by michaelhkay** [comment]
I'm now having great difficulty following the thread, since there seem to be comments on proposals that have since been changed. I think it would be clearer to close the issue and start a new one.

**Comment by ChristianGruen** [comment]
Closed; see #755 for further discussion.
